In [ ]:
### Dependencies
import pandas as pd
import numpy as np


## File Paths
# Path to mhsaa tables to merge

enrollment_path = 'data\school_info\mhsaa_enrolment_2022.csv'
name_color_path = 'data\school_info\mhsaa_school_nickname_color_2020.csv'

df_enrol = pd.read_csv(enrollment_path)
df_name_color = pd.read_csv(name_color_path)





In [ ]:
### 5-9-23

## Code to scrape 2023 MHSAA Tourny Data




In [ ]:
import pandas as pd

# Replace this with the plain text containing the tournament information
# read a text file into the variable text

text = open('2023_districts_raw.txt', 'r').read()

# Split the text into sections for each division
sections = text.split('Division ')

# Remove the first empty string
sections.pop(0)

# Initialize empty lists for each column in the dataframe
divisions = []
districts = []
hosts = []
locations = []

# Loop through the sections and extract the relevant information
for section in sections:
    lines = section.split('\n')
    division = 'Division ' + lines[0]
    for line in lines[1:]:
        if 'Baseball District' in line:
            district = line
        elif 'Host:' in line:
            host = line.split(': ')[1]
        elif 'Location:' in line:
            location = line.split(': ')[1]
        elif line != '':
            # Skip any blank lines
            districts.append(district)
            divisions.append(division)
            hosts.append(host)
            locations.append(location)

# Create a dataframe to store the extracted information
df = pd.DataFrame({'Division': divisions, 'District': districts, 'Host': hosts, 'Location': locations})

# Print the dataframe
print(df)


In [ ]:
### DF came back as a ton of duplicates. I need to clean it up.

df = df.drop_duplicates()

# Reindex the dataframe
df = df.reset_index(drop=True)

# output the dataframe to a csv file
df.to_csv('data/2023_district_hosts.csv', index=False)


print(df)

In [ ]:
df.head(20)

df.info()

## Regional Locations

In [ ]:
### URLS of pages with retional data

urls = {'Division 1': 'https://my.mhsaa.com/Sports/MHSAA-Tournament-Brackets/BracketGroup/9/Classification/1/SportSeasonId/424201',
        'Division 2': 'https://my.mhsaa.com/Sports/MHSAA-Tournament-Brackets/BracketGroup/9/Classification/2/SportSeasonId/424201',
        'Division 3': 'https://my.mhsaa.com/Sports/MHSAA-Tournament-Brackets/BracketGroup/9/Classification/3/SportSeasonId/424201',
        'Division 4': 'https://my.mhsaa.com/Sports/MHSAA-Tournament-Brackets/BracketGroup/9/Classification/4/SportSeasonId/424201'}

In [ ]:
import requests
from lxml import html
import pandas as pd

# Define the URLs for each division
urls = {'Division 1': 'https://my.mhsaa.com/Sports/MHSAA-Tournament-Brackets/BracketGroup/9/Classification/1/SportSeasonId/424201',
        'Division 2': 'https://my.mhsaa.com/Sports/MHSAA-Tournament-Brackets/BracketGroup/9/Classification/2/SportSeasonId/424201',
        'Division 3': 'https://my.mhsaa.com/Sports/MHSAA-Tournament-Brackets/BracketGroup/9/Classification/3/SportSeasonId/424201',
        'Division 4': 'https://my.mhsaa.com/Sports/MHSAA-Tournament-Brackets/BracketGroup/9/Classification/4/SportSeasonId/424201'}

# Initialize empty lists for each column in the dataframe
divisions = []
locations = []
links = []

# Loop through each division URL in the dictionary
for division, url in urls.items():
    try:
        # Send a GET request to the URL and parse the HTML content
        page = requests.get(url)
        tree = html.fromstring(page.content)

        # Find all the contest location spans using XPath
        location_spans = tree.xpath('//span[@class="contestlocation"]')

        # Loop through the location spans and extract the relevant information
        for location_span in location_spans:
            # Extract the location and link from the contest location span
            location = location_span.xpath('text()')[0].strip()
            link = location_span.xpath('a/@href')[0]
            # Append the information to the respective lists
            divisions.append(division)
            locations.append(location)
            links.append(link)
    except:
        print(f'Error: Failed to retrieve data for {division}')

# Create a dataframe to store the extracted information
df = pd.DataFrame({'Division': divisions, 'Location': locations, 'Link': links})

# Reset the index of the dataframe
df = df.reset_index(drop=True)

# Print the dataframe
print(df)


In [ ]:
## Clean up the dataframe

## Drop duplicates
df = df.drop_duplicates()

df.head(30)
# df.info()

In [ ]:
df.info()

### Output as a csv
## Might want to go back and adjust code to try to store which specific games are at each location
## regional has (semis and finals) then there is a quarterfinals round

df.to_csv('data/2023_regional_hosts.csv', index=False)

# Next Project

### Create a json with just the fields in michigan and try to integrate a column that marks the appropriate fields as host of districts and regionals

The text of the locations in the playoff csvs is not going to match the field names all that well. it might be worth trying to identify them from the map location - will have to go back to districts and extract map locations

In [ ]:
### Try to get all the google maps link from the districts page

import pandas as pd
from bs4 import BeautifulSoup

## Read local file
path = 'districts_2023.html'
html = open(path, 'r').read()


# Parse the HTML
soup = BeautifulSoup(html, 'html.parser')

# Find all the tournament divs
tournaments = soup.select('div.keep-together')

# Initialize lists to store data
district_numbers = []
hosts = []
locations = []

# Extract data for each tournament
for tournament in tournaments:
    district_number = tournament.find('span', {'data-bind': 'text:Division'}).text
    host = tournament.find('span', {'data-bind': 'text:Host'}).text
    location = tournament.find('a', {'target': '_blank'}).get('href')
    

    district_numbers.append(district_number)
    hosts.append(host)
    locations.append(location)

# Create a DataFrame
data = {'Division': district_numbers, 'Host': hosts, 'Location': locations}
df = pd.DataFrame(data)

# add another column the is the index number of the row + 1
df['District'] = df.index + 1

# Print the DataFrame
print(df)

df.to_csv('district_tournaments.csv', index=False)




# Try to match up locations from the scraped district and regional and link them to a field in my json data

## Stragegy: The district csv contains a field that has a link to a google maps search. Loop through all of those and return the lat and longitude coordinates then match the coordinates to the nearest home plate coordinate in the json

In [1]:
import pandas as pd
import googlemaps

## paths

local_json = 'data\michigan_fields.json'

district_csv = 'district_tournaments.csv'

regional_csv = 'data\2023_regional_hosts.csv'

# Replace this with your own API key
api_key = "AIzaSyA_BhlTupRdBPBhRptQuR6pYorMVYQnRMA"



In [2]:
# load Data
df = pd.read_csv(district_csv)

# df.head()

df.info()





<class 'pandas.core.frame.DataFrame'>
RangeIndex: 128 entries, 0 to 127
Data columns (total 4 columns):
 #   Column    Non-Null Count  Dtype 
---  ------    --------------  ----- 
 0   Division  128 non-null    int64 
 1   Host      128 non-null    object
 2   Location  126 non-null    object
 3   District  128 non-null    int64 
dtypes: int64(2), object(2)
memory usage: 4.1+ KB


In [3]:
### Clean up the location column to remove first portion and just leave the address remaining



# Remove the unwanted portion of the string in the 'Location' column
prefix = "http://maps.google.com/maps?q="

# Check if the location is a string before applying lstrip
df['Location'] = df['Location'].apply(lambda x: x.lstrip(prefix) if isinstance(x, str) else x)

# Print the cleaned DataFrame
print(df)

     Division                                Host  \
0           1                           Marquette   
1           1                         Midland Dow   
2           1                Muskegon Mona Shores   
3           1  Grand Rapids Forest Hills Northern   
4           1                          Grandville   
..        ...                                 ...   
123         4                             Genesee   
124         4  Sterling Heights Parkway Christian   
125         4     Waterford Our Lady of the Lakes   
126         4           Riverview Gabriel Richard   
127         4          Plymouth Christian Academy   

                                              Location  District  
0             North Marquette Fields,   Marquette, MI          1  
1    H H Dow High School - Baseball, 3901 N. Sagina...         2  
2    Mona Shores Baseball Field, 1121 W. Seminole R...         3  
3    FHN Stadium - Baseball, 3801 Leonard NE  Grand...         4  
4    Grandville High School 

In [4]:
df.head()

Division                                Host  \
0         1                           Marquette   
1         1                         Midland Dow   
2         1                Muskegon Mona Shores   
3         1  Grand Rapids Forest Hills Northern   
4         1                          Grandville   

                                            Location  District  
0           North Marquette Fields,   Marquette, MI          1  
1  H H Dow High School - Baseball, 3901 N. Sagina...         2  
2  Mona Shores Baseball Field, 1121 W. Seminole R...         3  
3  FHN Stadium - Baseball, 3801 Leonard NE  Grand...         4  
4  Grandville High School - Baseball, 4700 Canal ...         5

In [5]:
## USES GOOGLE CODE
import pandas as pd
import googlemaps

# Replace 'your_api_key' with your actual Google Maps API key
api_key = 'AIzaSyA_BhlTupRdBPBhRptQuR6pYorMVYQnRMA'
gmaps = googlemaps.Client(key=api_key)

# # Create a DataFrame from your data (use your actual DataFrame here)
# data = {
#     "Division": [1, 1, 1],
#     "Host": ["Marquette", "Midland Dow", "Muskegon Mona Shores"],
#     "Location": [
#         "North Marquette Fields, Marquette, MI",
#         "H H Dow High School - Baseball, 3901 N. Saginaw Rd. Midland, MI",
#         "Mona Shores Baseball Field, 1121 W. Seminole Rd. Muskegon, MI",
#     ],
#     "District": [1, 2, 3],
# }

# df = pd.DataFrame(data)

# Function to get the coordinates for a given address
def get_coordinates(address):
    geocode_result = gmaps.geocode(address)
    if geocode_result:
        lat = geocode_result[0]["geometry"]["location"]["lat"]
        lng = geocode_result[0]["geometry"]["location"]["lng"]
        return (lat, lng)
    else:
        return None

# Apply the function to the 'Location' column and store the coordinates in a new column
df["Coordinates"] = df["Location"].apply(get_coordinates)

# Print the updated DataFrame
print(df)


     Division                                Host  \
0           1                           Marquette   
1           1                         Midland Dow   
2           1                Muskegon Mona Shores   
3           1  Grand Rapids Forest Hills Northern   
4           1                          Grandville   
..        ...                                 ...   
123         4                             Genesee   
124         4  Sterling Heights Parkway Christian   
125         4     Waterford Our Lady of the Lakes   
126         4           Riverview Gabriel Richard   
127         4          Plymouth Christian Academy   

                                              Location  District  \
0             North Marquette Fields,   Marquette, MI          1   
1    H H Dow High School - Baseball, 3901 N. Sagina...         2   
2    Mona Shores Baseball Field, 1121 W. Seminole R...         3   
3    FHN Stadium - Baseball, 3801 Leonard NE  Grand...         4   
4    Grandville High Sc

In [6]:
## Check Output

df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 128 entries, 0 to 127
Data columns (total 5 columns):
 #   Column       Non-Null Count  Dtype 
---  ------       --------------  ----- 
 0   Division     128 non-null    int64 
 1   Host         128 non-null    object
 2   Location     126 non-null    object
 3   District     128 non-null    int64 
 4   Coordinates  126 non-null    object
dtypes: int64(2), object(3)
memory usage: 5.1+ KB


## Google code worked OK - returned coords for 126 of 128

### Below I am going to try to match up those coordinates to the michigan fields jsons

In [7]:
#### set up paths and load data(copied from above)


import pandas as pd
import json
# import googlemaps

## paths

local_json = 'data\michigan_fields.json'

district_csv = 'district_tournaments.csv'

regional_csv = 'data\2023_regional_hosts.csv'

# Replace this with your own API key
api_key = "AIzaSyA_BhlTupRdBPBhRptQuR6pYorMVYQnRMA"

# load Data
df = pd.read_csv(district_csv)

## Load MI fields data from json file


# Read the JSON file
with open(local_json) as f:
    data = json.load(f)

# Create a DataFrame from the JSON data
mi_df = pd.DataFrame(data)

mi_df.head()

# df.head()

park_name  \
0               Soo High School   
1     Frankfort-Elberta Area HS   
2                   Ironwood HS   
3  Onway Little League Field 2    
4                    Morenci HS   

                                                foul  \
0  [[-84.36583, 46.4813183], [-84.3658206, 46.482...   
1  [[-86.2258505, 44.6344082], [-86.2258357, 44.6...   
2  [[-90.1514578, 46.457154], [-90.1505766, 46.45...   
3  [[-84.2327091, 45.3607011], [-84.2328888, 45.3...   
4  [[-84.2069811, 41.7223814], [-84.2058371, 41.7...   

                                                 fop        level  \
0  [[-84.36583, 46.4813183], [-84.3645398, 46.481...  High School   
1  [[-86.2258505, 44.6344082], [-86.2270454, 44.6...  High School   
2  [[-90.1514578, 46.457154], [-90.1523415, 46.45...  High School   
3  [[-84.2327091, 45.3607011], [-84.2319152, 45.3...        Youth   
4  [[-84.2069811, 41.7223814], [-84.2071112, 41.7...  High School   

                  home_plate  foul_area_sqft  fop_area_sqft  field_area_sqft  \
0    [-84.36583, 46.4813183]    34451.384825   94783.943544    129235.328369   
1  [-86.2258505, 44.6344082]    25687.658559   86823.965374    112511.623933   
2   [-90.1514578, 46.457154]    37066.246977   95913.228498    132979.475475   
3  [-84.2327091, 45.3607011]    10384.795715   35156.473366     45541.269081   
4  [-84.2069811, 41.7223814]    35463.259931   92763.241003    128226.500934   

   foul_area_per  fair_to_foul  ...        country  school_id  school_name  \
0       0.266579      2.751238  ...  United States        NaN         None   
1       0.228311      3.379988  ...  United States        NaN         None   
2       0.278737      2.587616  ...  United States        NaN         None   
3       0.228030      3.385379  ...  United States        NaN         None   
4       0.276567      2.615756  ...  United States        NaN         None   

   students  division    Nickname   Color1  Color2  Color3  Color4  
0       NaN      None        None     None    None    None    None  
1       NaN      None        None     None    None    None    None  
2       NaN      None  Red Devils     Red    White    None    None  
3       NaN      None        None     None    None    None    None  
4       NaN      None    Bulldogs  Maroon    White    None    None  

[5 rows x 37 columns]

In [9]:
### Try a different approach to match the fields
## Use the Host name to find 3 matches from the mi_df
import pandas as pd
from fuzzywuzzy import fuzz, process

# Assuming you have the two dataframes df and mi_df

def find_closest_park_names(host, n_closest=3):
    closest_park_names = process.extract(host, mi_df["park_name"], limit=n_closest, scorer=fuzz.token_sort_ratio)
    return [name for name, score, index in closest_park_names]

# Apply the function to the 'Host' column and store the results in new columns
df[["closest_park_1", "closest_park_2", "closest_park_3"]] = df["Host"].apply(find_closest_park_names).apply(pd.Series)

# Print the updated DataFrame
print(df)


     Division                                Host  \
0           1                           Marquette   
1           1                         Midland Dow   
2           1                Muskegon Mona Shores   
3           1  Grand Rapids Forest Hills Northern   
4           1                          Grandville   
..        ...                                 ...   
123         4                             Genesee   
124         4  Sterling Heights Parkway Christian   
125         4     Waterford Our Lady of the Lakes   
126         4           Riverview Gabriel Richard   
127         4          Plymouth Christian Academy   

                                              Location  District  \
0    http://maps.google.com/maps?q=North Marquette ...         1   
1    http://maps.google.com/maps?q=H H Dow High Sch...         2   
2    http://maps.google.com/maps?q=Mona Shores Base...         3   
3    http://maps.google.com/maps?q=FHN Stadium - Ba...         4   
4    http://maps.google

In [11]:
### FUNCTION TO FIND NEAREST FIELD TO DISTRICT TOURNAMENT LOCATION

import math

## Define a function to calculate the Haversine distance between two points
def haversine_distance(coord1, coord2):
    # Convert latitude and longitude to radians
    lat1, lon1 = map(math.radians, coord1)
    lat2, lon2 = map(math.radians, coord2)

    # Calculate the differences between latitudes and longitudes
    dlat = lat2 - lat1
    dlon = lon2 - lon1

    # Calculate the Haversine distance
    a = math.sin(dlat / 2) ** 2 + math.cos(lat1) * math.cos(lat2) * math.sin(dlon / 2) ** 2
    c = 2 * math.asin(math.sqrt(a))
    r = 6371  # Radius of the Earth in km

    return c * r

def find_closest_parks(coord, n_closest=3):
    if coord is None:
        return ["Unknown"] * n_closest

    mi_df["distance"] = mi_df["home_plate"].apply(lambda x: haversine_distance(coord, (x[1], x[0])))
    closest_park_indices = mi_df["distance"].nsmallest(n_closest).index
    return mi_df.loc[closest_park_indices, "park_name"].tolist()



# Make sure 'home_plate' in mi_df has coordinates in the format (lat, lng)
mi_df["home_plate"] = mi_df["home_plate"].apply(lambda x: (x[0], x[1]))

# Create a new column 'closest_park' in df
df[["closest_park_1", "closest_park_2", "closest_park_3"]] = df["Coordinates"].apply(find_closest_parks).apply(pd.Series)


# Print the updated DataFrame
print(df)



     Division                                Host  \
0           1                           Marquette   
1           1                         Midland Dow   
2           1                Muskegon Mona Shores   
3           1  Grand Rapids Forest Hills Northern   
4           1                          Grandville   
..        ...                                 ...   
123         4                             Genesee   
124         4  Sterling Heights Parkway Christian   
125         4     Waterford Our Lady of the Lakes   
126         4           Riverview Gabriel Richard   
127         4          Plymouth Christian Academy   

                                              Location  District  \
0             North Marquette Fields,   Marquette, MI          1   
1    H H Dow High School - Baseball, 3901 N. Sagina...         2   
2    Mona Shores Baseball Field, 1121 W. Seminole R...         3   
3    FHN Stadium - Baseball, 3801 Leonard NE  Grand...         4   
4    Grandville High Sc

In [10]:
df.head()

Division                                Host  \
0         1                           Marquette   
1         1                         Midland Dow   
2         1                Muskegon Mona Shores   
3         1  Grand Rapids Forest Hills Northern   
4         1                          Grandville   

                                            Location  District  \
0  http://maps.google.com/maps?q=North Marquette ...         1   
1  http://maps.google.com/maps?q=H H Dow High Sch...         2   
2  http://maps.google.com/maps?q=Mona Shores Base...         3   
3  http://maps.google.com/maps?q=FHN Stadium - Ba...         4   
4  http://maps.google.com/maps?q=Grandville High ...         5   

                          closest_park_1  \
0                        MarquetteField    
1                 Dow Diamond (Midland)    
2                Muskegon Mona Shores HS   
3  Grand Rapids Forest Hills Northern HS   
4                          Greenville HS   

                                     closest_park_2  \
0                            Marquette HS - Field 2   
1                                          Milan HS   
2                              Muskegon Oakridge HS   
3  Grand Rapids Forest Hills Northern HS - practice   
4                                      Dansville HS   

               closest_park_3  
0  Marquette HS - Haley Field  
1                    Dowagiac  
2           North Muskegon HS  
3   Grand Rapids Northview HS  
4        Greenville JV / 16u

In [11]:
## Output this matching as a csv so I can manulauly check it

df.to_csv('district_fields_text_match.csv', index=False)

## End of 5-9-23 Work for now. output csv file with possible matches for the district fields


# Working with outlier fields 

## Start 59/23 Night


In [1]:
## Dependencies and Setup
import matplotlib.pyplot as plt
import pandas as pd
import scipy.stats as st
import numpy as np

# Load files
df_out = pd.read_csv('outlier_fields.csv')

df_out.head()


park_name  \
0                            Manton HS   
1  Adrian Lenawee Christian HS- replot   
2              Ann Arbor Greenhills HS   
3                 Ann Arbor Skyline HS   
4                          Bellaire HS   

                                                foul  \
0  [(-85.4053602, 44.4128512), (-85.4041559, 44.4...   
1  [(-84.08374923296931, 41.90876073773172), (-84...   
2  [(-84.01479353123523, 41.86694552754826), (-84...   
3  [(-83.77545696328717, 42.30415973972723), (-83...   
4  [(-85.19125608380878, 44.97663506300876), (-85...   

                                                 fop        level  \
0  [(-85.4053602, 44.4128512), (-85.4052235, 44.4...  High School   
1  [(-84.08263859810114, 41.90903328830586), (-84...  High School   
2  [(-84.01374704390223, 41.86720354412398), (-84...  High School   
3  [(-83.77427962289079, 42.30418142498592), (-83...  High School   
4  [(-85.19062986396423, 44.97594663020872), (-85...  High School   

                                home_plate  foul_area_sqft  fop_area_sqft  \
0                (-85.4053602, 44.4128512)    81352.834403   20266.747151   
1  (-84.08263859810114, 41.90903328830586)    23497.955671   87428.696853   
2  (-84.01374704390223, 41.86720354412398)    39865.290607   80393.426329   
3  (-83.77427962289079, 42.30418142498592)    32529.386734   94610.656524   
4  (-85.19062986396423, 44.97594663020872)    22443.107113   85074.553380   

   field_area_sqft  foul_area_per  fair_to_foul  ...        country  \
0    101619.581554       0.800563      0.249122  ...  United States   
1    110926.652524       0.211833      3.720694  ...  United States   
2    120258.716936       0.331496      2.016627  ...  United States   
3    127140.043258       0.255855      2.908467  ...  United States   
4    107517.660494       0.208739      3.790676  ...  United States   

   school_id                school_name  students  division  Nickname  \
0     7087.0                  Manton HS     275.0         C   Rangers   
1        NaN                        NaN       NaN       NaN       NaN   
2     6543.0  Ann Arbor - Greenhills HS     319.0         C  Gryphons   
3     2032.0     Ann Arbor - Skyline HS    1588.0         A    Eagles   
4     2724.0                Bellaire HS     101.0         D    Eagles   

           Color1   Color2  Color3  Color4  
0         Orange     Black     NaN     NaN  
1             NaN      NaN     NaN     NaN  
2           Navy    Forest     NaN     NaN  
3  Columbia Blue     White     NaN     NaN  
4            Red     Black     NaN     NaN  

[5 rows x 37 columns]

In [3]:
## Find the actual homeplate for each field

## Look through the 'fop' coulmn and find any points that are in the list twice
## The repeated points should be the homeplate

## Loop through the fop column and find the repeated points
## Add the repeated points to a list with the field name

## Create a new column in the df_out dataframe with the homeplate coordinates
for each in df_out['fop']:
    if each in df_out['fop']:
        print(each)
    else:
        print('no match')

## Create a new column in the df_out dataframe with the homeplate coordinates
df_out['homeplate_extract'] = df_out['fop'].str.extract(r'(\d{2}\.\d{4,5})')


IndentationError: unexpected indent (1586301994.py, line 10)

In [ ]:
import pandas as pd
import requests
from lxml import html

url = 'https://my.mhsaa.com/Sports/Baseball/Districts'  # Replace this with the URL of the webpage you want to scrape
page = requests.get(url)
tree = html.fromstring(page.content)

# Find the game location using the XPath
location = tree.xpath('/html/body/form/div[5]/div[2]/div/div/div[2]/div[1]/div/div/div/div/div/div/div/div[2]/div[2]/div[1]/div[2]/div[1]/a')[0]

# Extract the relevant information
name = location.text.strip()
link = location.get('href')
address = link.split('=')[1].strip()

# Create a dataframe to store the scraped data
df = pd.DataFrame({'Field_name': [name], 'Location': [address], 'Link': [link]})

# Print the dataframe
print(df)


# Get latitude and longitude for each location
df["Coordinates"] = df["Location"].apply(get_latitude_longitude)

# Print the DataFrame
print(df)




In [ ]:
print(df_name_color.columns)

In [ ]:
### Match the school name to School and merge the dataframes into a single object
from fuzzywuzzy import fuzz, process

def find_best_match(school_name, choices, score_cutoff=70):
    best_match = process.extractOne(school_name, choices, scorer=fuzz.token_sort_ratio, score_cutoff=score_cutoff)
    if best_match:
        return best_match[0]
    else:
        return None

# Get the list of school names from df_name_color
school_names = df_name_color['School'].tolist()

# Apply find_best_match function to create a new column 'best_match' in df_enrol
df_enrol['best_match'] = df_enrol['school_name'].apply(find_best_match, choices=school_names, score_cutoff=80)



In [ ]:
# Rename the 'School' column in df_name_color to 'best_match'
df_name_color = df_name_color.rename(columns={'School': 'best_match'})

# Merge the dataframes on the 'best_match' column
df_merged = df_enrol.merge(df_name_color, on='best_match', how='left')


In [ ]:
print(df_enrol.columns)

print(df_name_color.columns)

In [ ]:
# - get icons sorted out
# -*** DONE*** get the level assigner sorted out in the etl
# - add filter based on level to map
# - implement the search box places from 